server/src/com/cloud/resource/ResourceManagerImpl.java

```diff
@@ -1296,10 +1296,17 @@ public boolean checkAndMaintain(final long hostId) {
             if (host.getType() != Host.Type.Storage) {
                 final List<VMInstanceVO> vos = _vmDao.listByHostId(hostId);
                 final List<VMInstanceVO> vosMigrating = _vmDao.listVmsMigratingFromHost(hostId);
+                final List<VMInstanceVO> failedMigratedVms = _vmDao.listNonMigratingVmsByHostEqualsLastHost(hostId);
                 if (vos.isEmpty() && vosMigrating.isEmpty()) {
-                    resourceStateTransitTo(host, ResourceState.Event.InternalEnterMaintenance, _nodeId);
-                    hostInMaintenance = true;
-                    ActionEventUtils.onCompletedActionEvent(CallContext.current().getCallingUserId(), CallContext.current().getCallingAccountId(), EventVO.LEVEL_INFO, EventTypes.EVENT_MAINTENANCE_PREPARE, "completed maintenance for host " + hostId, 0);
+                    if (!failedMigratedVms.isEmpty()) {
```

Comment:
@nvazquez, I think if we put this condition in line 1300 itself, then it would maintain the old flow. Something like this in line 1300: if (vos.isEmpty() && vosMigrating.isEmpty() && failedMigratedVms.isEmpty()) {...}. And we can put this message for all kind of failures, where resource state would transit to UnableToMigrate.